### ISSUE TITLE
[RFC]: support customized scheduler & pipelines for diffusion model with arbitrary returns

### ISSUE THREAD

**Issue by zhtmike** [vllm-project/vllm-omni#686]
### Motivation.

Currently, vllm-omni is limited to returning images and videos only through predefined pipelines and schedulers.
In reinforcement learning (RL) scenarios, however, we often use customized schedulers (such as the SDE version of the Euler sampler) instead of the original one during the rollout stage. We also need access to intermediate values stored in the denoising stages on the inference side, such as prompt embeddings and cached latents for the training loop. Without access to this information from inference engine, we are forced to recompute these values in the training engine repeatedly, leading to redundant computations.

### Proposed Change.

Provide an interface that allows users to create their own schedulers or pipelines, and supports arbitrary outputs from user-defined pipelines.

Changes in vllm-omni to support customized pipeline (by @knlnguyen1802 ): 

![Image](https://github.com/user-attachments/assets/d3f313da-7739-46cb-8939-8c521fa02eec)
 
The customized pipeline can be called from RPC (#371) with the following interface

```python
from vllm_omni.diffusion.data import DiffusionOutput, OmniDiffusionConfig
from vllm_omni.diffusion.request import OmniDiffusionRequest

class CustomPipeline:
    def __init__(self, od_config: OmniDiffusionConfig):
        pass

    def forward(self, request: OmniDiffusionRequest) -> DiffusionOutput:
        pass

    def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
        pass

    def setup_tranformer_engine(self, ...):
        self.tranformer = ...
```

### CC List.

@wtomin @Gaohan123 @SamitHuang 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by wtomin** [comment]
It is true that current `DiffusionOutput` has placeholders for extra outputs, like `trajectory_latents`

```python

@dataclass
class DiffusionOutput:
    """
    Final output (after pipeline completion)
    """

    output: torch.Tensor | None = None
    trajectory_timesteps: list[torch.Tensor] | None = None
    trajectory_latents: torch.Tensor | None = None
    trajectory_decoded: list[torch.Tensor] | None = None
    error: str | None = None

    post_process_func: Callable[..., Any] | None = None

```

But in actual diffusion models pipelines, only the final denoised latent or the final decoded image will be returned, for example:
```python
class StableDiffusion3Pipeline(
    nn.Module,
):
    def forward(
        ... ...
        if self.output_type == "latent":
            image = latents
        else:
            latents = latents.to(self.vae.dtype)
            latents = (latents / self.vae.config.scaling_factor) + self.vae.config.shift_factor

            image = self.vae.decode(latents, return_dict=False)[0]

        return DiffusionOutput(output=image)
```

What do you think would be the best interface to allow users to customize the returned tensors?

**Comment by zhtmike** [comment]
A generalized class interface should be sufficient:

```python
class UserDefinedPipeline(ABC):
    @abstractmethod 
    def forward(...): 
        return ArbitrarilyOutput(output=image, encoder_state=encoder_state, ...)
```

Suppose I have implemented a custom pipeline based on this abstraction. How can I inject this pipeline into vllm-omni without modifying its source code, while ensuring that it works properly just like the built-in pipelines?

**Comment by zhtmike** [comment]
@knlnguyen1802 :)

**Comment by hsliuustc0106** [comment]
i think this a big feature for vllm-omni, please prepare a detailed design document for this feature before implementation.

**Comment by knlnguyen1802** [comment]
FYI: vLLM allow to custom and extension worker 
https://github.com/vllm-project/vllm/blob/0d7667419f738e44ce7f9bf311987e15b01970b0/vllm/v1/worker/worker_base.py#L257

**Comment by zhtmike** [comment]
And some reference:
https://github.com/huggingface/diffusers/blob/main/docs/source/en/using-diffusers/custom_pipeline_overview.md

@AndyZhou952 

The basic paradigm is to write a custom `custom_pipeline.py`, which allows `vllm-omni` to automatically load and register the user-defined pipeline by some methods

**Comment by zhtmike** [comment]
> FYI: vLLM allow to custom and extension worker https://github.com/vllm-project/vllm/blob/0d7667419f738e44ce7f9bf311987e15b01970b0/vllm/v1/worker/worker_base.py#L257

A custom worker may be too low-level for my concern. We still need features that are already implemented in the existing diffusion.worker.

**Comment by knlnguyen1802** [comment]
> > FYI: vLLM allow to custom and extension worker https://github.com/vllm-project/vllm/blob/0d7667419f738e44ce7f9bf311987e15b01970b0/vllm/v1/worker/worker_base.py#L257
> 
> A custom worker may be too low-level for my concern. We still need features that are already implemented in the existing diffusion.worker.

It's only the worker extension, not rewrite the whole worker
For the case of custom pipeline, it can be implemented by reuse the `WorkerExtension` in this way

``` python
class WorkerExtension:
    def re_init_pipeline(pipeline: CustomPipeLine):
          pass

omniDiffusion = Omni(...,worker_extension_cls=WorkerExtension)

```

This might need vllm-Omni to support option to init dummy pipeline and it is necessary to rewrite for compatible.

There is example work in this way for vLLM: 
https://github.com/vllm-project/vllm/blob/main/examples/offline_inference/rlhf_colocate.py 
https://github.com/vllm-project/vllm/blob/main/examples/offline_inference/rlhf_utils.py

**Comment by zhtmike** [comment]
> > > FYI: vLLM allow to custom and extension worker https://github.com/vllm-project/vllm/blob/0d7667419f738e44ce7f9bf311987e15b01970b0/vllm/v1/worker/worker_base.py#L257
> > 
> > 
> > A custom worker may be too low-level for my concern. We still need features that are already implemented in the existing diffusion.worker.
> 
> It's only the worker extension, not rewrite the whole worker For the case of custom pipeline, it can be implemented by reuse the `WorkerExtension` in this way
> 
> class WorkerExtension:
>     def re_init_pipeline(pipeline: CustomPipeLine):
>           pass
> 
> omniDiffusion = Omni(...,worker_extension_cls=WorkerExtension)
> This might need vllm-Omni to support option to init dummy pipeline and it is necessary to rewrite for compatible.
> 
> There is example work in this way for vLLM: https://github.com/vllm-project/vllm/blob/main/examples/offline_inference/rlhf_colocate.py https://github.com/vllm-project/vllm/blob/main/examples/offline_inference/rlhf_utils.py

It seems plausible, FYI @AndyZhou952

**Comment by zhtmike** [comment]
@chenyingshu pls take a look~

**Comment by hsliuustc0106** [comment]
@natureofnature do we need to add a modelrunner under worker?

**Comment by ZJY0516** [comment]
@knlnguyen1802 since we have several RL related issues, could you please open a new issue to track them?

**Comment by knlnguyen1802** [comment]
> [@knlnguyen1802](https://github.com/knlnguyen1802) since we have several RL related issues, could you please open a new issue to track them?

Ok I will do it
